### PR TITLE
Fleet mode and Fleet Server support in Elastic Agent CRD

### PR DESCRIPTION
**Pull request by david-kow**
This PR introduces support for Fleet mode and Fleet Server in our Agent CRD. Closes https://github.com/elastic/cloud-on-k8s/issues/4429.

#### New fields

```yaml
mode: fleet # or standalone
fleetServerEnabled: true
http: ... # configuration for Fleet Server endpoint
kibanaRef: ... # allows to setup Fleet and enroll
fleetServerRef: ... # allows to connect to Fleet Server automatically
```

#### Basic setup

To run Fleet Server, user is expected to set:
```yaml
mode: fleet
elasticsearchRefs:
- name: elasticsearch
fleetServerEnabled: true
kibanaRef: kibana
http: ... # optionally, as needed
```

Tu run Elastic Agent in Fleet mode, user is expected to set:
```yaml
mode: fleet
kibanaRef: kibana
fleetServerRef: fleet-server
```

#### Examples to try

<details>
<summary>Elasticsearch, Kibana, Fleet Server, Elastic Agent in ECK</summary>

```yaml
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: kibana
spec:
  version: 7.14.0-SNAPSHOT
  count: 1
  elasticsearchRef:
    name: elasticsearch
  config:
    xpack.fleet.agents.elasticsearch.host: https://elasticsearch-es-http.default.svc:9200
    xpack.fleet.agents.fleet_server.hosts: ["https://fleet-server-agent-http.default.svc:8220"]
---
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  version: 7.14.0-SNAPSHOT
  nodeSets:
  - name: default
    count: 3
    config:
      node.store.allow_mmap: false
    volumeClaimTemplates:
      - metadata:
          name: elasticsearch-data
        spec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 50Gi
---
apiVersion: agent.k8s.elastic.co/v1alpha1
kind: Agent
metadata:
  name: fleet-server
spec:
  version: 7.14.0-SNAPSHOT
  kibanaRef:
    name: kibana
  elasticsearchRefs:
  - name: elasticsearch
  mode: fleet
  fleetServerEnabled: true
  deployment:
    replicas: 1
    podTemplate:
      spec:
        automountServiceAccountToken: true
        securityContext:
          runAsUser: 0
        serviceAccountName: fleet-server
---
apiVersion: agent.k8s.elastic.co/v1alpha1
kind: Agent
metadata: 
  name: elastic-agent
  labels:
    app: app8
spec:
  version: 7.14.0-SNAPSHOT
  kibanaRef:
    name: kibana
  fleetServerRef: 
    name: fleet-server
  mode: fleet
  daemonSet:
    podTemplate:
      spec:
        automountServiceAccountToken: true
        serviceAccountName: elastic-agent
        securityContext:
          runAsUser: 0
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: fleet-server
rules:
- apiGroups: [""] # "" indicates the core API group
  resources:
  - namespaces
  - pods
  verbs:
  - get
  - watch
  - list
- apiGroups: ["coordination.k8s.io"]
  resources:
  - leases
  verbs:
  - get
  - create
  - update
- nonResourceURLs:
  - "/metrics"
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - ""
  resources:
  - namespaces
  - pods
  - nodes
  - nodes/metrics
  - nodes/proxy
  - nodes/stats
  - events
  verbs:
  - get
  - watch
  - list
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: fleet-server
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: fleet-server
subjects:
- kind: ServiceAccount
  name: fleet-server
  namespace: default
roleRef:
  kind: ClusterRole
  name: fleet-server
  apiGroup: rbac.authorization.k8s.io
---
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    name: elastic-agent
  rules:
  - apiGroups: [""] # "" indicates the core API group
    resources:
    - namespaces
    - pods
    verbs:
    - get
    - watch
    - list
  - apiGroups: ["coordination.k8s.io"]
    resources:
    - leases
    verbs:
    - get
    - create
    - update
  - nonResourceURLs:
    - "/metrics"
    verbs:
    - get
    - watch
    - list
  - apiGroups:
    - ""
    resources:
    - nodes
    - nodes/metrics
    - nodes/proxy
    - nodes/stats
    - events
    verbs:
    - get
    - watch
    - list
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elastic-agent
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: elastic-agent
subjects:
- kind: ServiceAccount
  name: elastic-agent
  namespace: default
roleRef:
  kind: ClusterRole
  name: elastic-agent
  apiGroup: rbac.authorization.k8s.io
---
```

</details>

```
$ k tree agents fleet-server-test
NAMESPACE  NAME                                                               READY  REASON  AGE
default    Agent/fleet-server-test                                            -              6m46s
default    ├─Deployment/fleet-server-test-agent                               -              6m39s
default    │ └─ReplicaSet/fleet-server-test-agent-589cd4dd87                  -              6m39s
default    │   └─Pod/fleet-server-test-agent-589cd4dd87-cm6g8                 True           6m38s
default    ├─Secret/fleet-server-test-agent-config                            -              6m39s
default    ├─Secret/fleet-server-test-agent-es-default-elasticsearch-test-ca  -              6m44s
default    ├─Secret/fleet-server-test-agent-http-ca-internal                  -              6m39s
default    ├─Secret/fleet-server-test-agent-http-certs-internal               -              6m39s
default    ├─Secret/fleet-server-test-agent-kb-user                           -              6m43s
default    ├─Secret/fleet-server-test-agent-kibana-ca                         -              6m43s
default    ├─Secret/fleet-server-test-default-elasticsearch-test-agent-user   -              6m44s
default    └─Service/fleet-server-test-agent-http                             -              6m39s
default      └─EndpointSlice/fleet-server-test-agent-http-chsfh               -              6m39s
$ k tree agents elastic-agent-test
NAMESPACE  NAME                                                        READY  REASON  AGE
default    Agent/elastic-agent-test                                    -              6m54s
default    ├─DaemonSet/elastic-agent-test-agent                        -              6m46s
default    │ ├─ControllerRevision/elastic-agent-test-agent-86d56fc59b  -              6m46s
default    │ ├─Pod/elastic-agent-test-agent-65pgt                      True           6m46s
default    │ ├─Pod/elastic-agent-test-agent-q2hjj                      True           6m46s
default    │ └─Pod/elastic-agent-test-agent-q7xpc                      True           6m46s
default    ├─Secret/elastic-agent-test-agent-config                    -              6m46s
default    ├─Secret/elastic-agent-test-agent-fleetserver-ca            -              6m54s
default    ├─Secret/elastic-agent-test-agent-kb-user                   -              6m51s
default    └─Secret/elastic-agent-test-agent-kibana-ca                 -              6m51s
```

Screenshot 2021-07-05 at 16 28 40

<details>
<summary>Elastic Agent only in ECK; Elasticsearch, Kibana and Fleet Server in Elastic Cloud</summary>

Change all `changeme` entries to appropriate values:
- `KIBANA_HOST` is the Elastic Cloud Kibana endpoint, can be found in the user console
- `FLEET_URL` is the Elastic Cloud Fleet endpoint, can be found in the user console
- `KIBANA_USERNAME`, `KIBANA_PASSWORD` could be the `elastic` user credentials

```yaml
apiVersion: agent.k8s.elastic.co/v1alpha1
kind: Agent
metadata: 
  name: elastic-agent
spec:
  version: 7.13.2
  mode: fleet
  daemonSet:
    podTemplate:
      spec:
        automountServiceAccountToken: true
        serviceAccountName: elastic-agent
        containers:
        - name: agent
          env:
          - name: KIBANA_HOST
            value: changeme
          - name: KIBANA_USERNAME
            value: changeme
          - name: KIBANA_PASSWORD
            value: changeme
          - name: FLEET_URL
            value: changeme
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elastic-agent
rules:
- apiGroups: [""] # "" indicates the core API group
  resources:
  - namespaces
  - pods
  verbs:
  - get
  - watch
  - list
- apiGroups: ["coordination.k8s.io"]
  resources:
  - leases
  verbs:
  - get
  - create
  - update
- nonResourceURLs:
  - "/metrics"
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - ""
  resources:
  - nodes
  - nodes/metrics
  - nodes/proxy
  - nodes/stats
  - events
  verbs:
  - get
  - watch
  - list
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elastic-agent
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: elastic-agent
subjects:
- kind: ServiceAccount
  name: elastic-agent
  namespace: default
roleRef:
  kind: ClusterRole
  name: elastic-agent
  apiGroup: rbac.authorization.k8s.io
---
```

</details>

Screenshot 2021-07-05 at 16 21 53
Screenshot 2021-07-05 at 16 30 33

Not providing Elastic Agent, Fleet Server on ECK with ES/KB in Cloud example as `7.13` Elastic Agent bug requires some workarounds in the Pod template. Nevertheless, it is possible to set it up. Also, those issues are fixed for `7.14`.

#### Known limitations

1. Due to [configuration limitations](https://github.com/elastic/kibana/issues/102794) on Fleet/Elastic Agent side we need to establish trust to ES via call to `update-ca-trust`. This is temporary, but now it will require Pods to be run with elevated privileges. Otherwise, we can't make Beats that Elastic Agent manages to trust Elasticsearch CA.

2. Related to 1., we currently don't have a good mechanism in place to fetch transitive Secret resource (ie. given EA -> FS -> ES, we can't get ES CA Secret for EA). This is an issue only when EA and ES are in different namespaces. This will go away together with 1., when CA is pushed from Fleet and ECK is not involved.

#### Open issue

There are two settings that Kibana has to have to configure Fleet correctly:
```yaml
xpack.fleet.agents.elasticsearch.host: https://elasticsearch-es-http.default.svc:9200
xpack.fleet.agents.fleet_server.hosts: ["https://fleet-server-agent-http.default.svc:8220"]
```

This PR assumes that user is responsible for setting them up. The reason for that is that it's the first time where a reference _to_ a resource, would cause that resource to have configuration modified. That is, a reference from Fleet Server to Kibana would cause that Kibana to have its configuration changed. Today, we have no mechanisms to detect a reference on the target side. We also don't have a way to validate if such change is permitted (as we have for refs with `accessReviewer`). I think we could add it, but wanted to raise it before I make code changes. This is somewhat separate from the rest of the changes and if it needs to be addressed, it can be through a separate PR.

### CODE REVIEW

**Comment by david-kow** [comment]
run full pr build

File: pkg/controller/agent/pod.go [hunk 1]
```diff
@@ -50,63 +63,197 @@ var (
...
+	if params.Agent.Spec.Mode == "" || params.Agent.Spec.Mode == agentv1alpha1.AgentStandaloneMode {
+		builder = builder.
+			WithResources(defaultResources).
+			WithArgs("-e", "-c", ConfigMountPath)
+	} else if params.Agent.Spec.Mode == agentv1alpha1.AgentFleetMode {
+		builder = builder.
+			WithResources(defaultFleetResources).
+			// needed to pick up fleet-setup.yml correctly
+			WithEnv(corev1.EnvVar{Name: "CONFIG_PATH", Value: "/usr/share/elastic-agent"})
+	}
```
Comment (pebrc) [inline pkg/controller/agent/pod.go:208]: Would a switch on `.Mode` improve readability a little?

File: pkg/controller/agent/pod.go [hunk 2]
```diff
@@ -50,63 +63,197 @@ var (
...
+				// adding Fleet Server associations (which includes Elasticsearch) to config hash attached to Elastic Agent
+				// Pods.
+				if err := commonassociation.WriteAssocsToConfigHash(params.Client, fs.GetAssociations(), configHash); err != nil {
+					return corev1.PodTemplateSpec{}, err
+				}
+
+				builder = builder.
+					WithCommand([]string{"/usr/bin/env", "bash", "-c", trustCAScript(path.Join(certificatesDir(esfsAssoc), CAFileName))})
+			}
+		}
```
Comment (pebrc) [inline pkg/controller/agent/pod.go:170]: Same here I think it would help readability to break it down into smaller helper functions otherwise it is quite hard to follow the large `if/else` clauses.

**Review by pebrc — COMMENTED** [review]
Did a first pass. Overall looks good! Having one thing (Agent) do multiple things (standalone, fleet, fleet server)  complicates code paths of course, I left comment were I thought things were a bit hard to read. I tried the sample you provided and it worked (the `fleetServerEnabled` attribute name has changed though) 

As discussed offline there is some potential to automate the Kibana settings a bit more but I don't think it's a must for this first version of experimental Fleet support.

File: pkg/controller/agent/config.go [hunk 3]
```diff
@@ -120,3 +147,98 @@ func getUserConfig(params Params) (*settings.CanonicalConfig, error) {
...
+
+			fleetServerCfg["elasticsearch"] = map[string]interface{}{
+				"ca":       esCA,
+				"host":     esHost,
+				"username": esUsername,
+				"password": esPassword,
+			}
+		}
+
+		cfgMap["fleet_server"] = fleetServerCfg
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/agent/config.go:210]: I've added constants for keys used at least twice.

File: pkg/controller/agent/config.go [hunk 4]
```diff
@@ -120,3 +147,98 @@ func getUserConfig(params Params) (*settings.CanonicalConfig, error) {
...
+			},
+		}
+	}
+
+	//nolint:nestif
+	if spec.EnableFleetServer {
+		cfgMap["fleet"] = map[string]interface{}{
+			"enroll": true,
+			"ca":     path.Join(FleetCertMountPath, certificates.CAFileName),
+			"url":    fmt.Sprintf("https://%s.%s.svc:8220", HTTPServiceName(params.Agent.Name), params.Agent.Namespace),
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/agent/config.go]: Changed to call to `association.ServiceURL`.

File: pkg/controller/agent/driver.go [hunk 5]
```diff
@@ -23,6 +30,10 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/utils/log"
 )
 
+const (
+	FleetServerPort int32 = 8220
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/agent/driver.go:34]: It is configurable through an environment variable. IIUC, we don't allow this with other resources (Pod ports and service ports are hardcoded) - I'm making this consistent with that.

**Comment by sebgl** [comment]
Sorry for being very late to this.
I think the `mode: standalone | fleet_server | fleet_agent ` alternative you proposed in https://github.com/elastic/cloud-on-k8s/issues/4429 would make slightly more sense to me?

File: pkg/controller/agent/driver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/agent/driver.go:34]: I'm fine with that. We can reconsider if it ever becomes a problem.
Users can do port redirection through services, while we keep our own hardcoded port at the Pod level.

File: pkg/apis/agent/v1alpha1/agent_types.go [hunk 6]
```diff
@@ -66,6 +66,30 @@ type AgentSpec struct {
...
+
+	// KibanaRef is a reference to Kibana where Fleet should be set up and this Agent should be enrolled. Don't set
+	// unless `mode` is set to `fleet`.
+	// +kubebuilder:validation:Optional
+	KibanaRef commonv1.ObjectSelector `json:"kibanaRef,omitempty"`
+
+	// FleetServerRef is a reference to Fleet Server that this Agent should connect to to obtain it's configuration.
+	// Don't set unless `mode` is set to `fleet`.
+	// +kubebuilder:validation:Optional
+	FleetServerRef commonv1.ObjectSelector `json:"fleetServerRef,omitempty"`
```
Comment (sebgl) [inline pkg/apis/agent/v1alpha1/agent_types.go:92]: To better encapsulate fleet stuff I'm wondering if we could reorganize this into:

```
fleet:
    managed: true
    server: false
    serverRef: <>
    kibanaRef: <>
```

File: pkg/controller/agent/pod.go [hunk 7]
```diff
@@ -50,63 +63,199 @@ var (
...
+
+				caSecretName := esAssociation.AssociationConf().GetCASecretName()
+				vols = append(vols, volume.NewSecretVolumeWithMountPath(
+					caSecretName,
+					fmt.Sprintf("%s-certs", esAssociation.AssociationType()),
+					certificatesDir(esAssociation),
+				))
+
+				// Because of the reference chain (Elastic Agent ---> Fleet Server ---> Elasticsearch), we are going to get
+				// notified when CA of Elasticsearch changes as Fleet Server resource will get updated as well. But what we
```
Comment (sebgl) [inline pkg/controller/agent/pod.go:190]: not sure I understand, what will change if the CA changes? Fleet Server gets reconciled but why would the agent reconcile?

File: pkg/controller/agent/pod.go [hunk 8]
```diff
@@ -50,63 +63,199 @@ var (
...
+				// also need to do is to roll Elastic Agent Pods to pick up the update CA. To do be able to do that, we are
+				// adding Fleet Server associations (which includes Elasticsearch) to config hash attached to Elastic Agent
+				// Pods.
+				if err := commonassociation.WriteAssocsToConfigHash(params.Client, fs.GetAssociations(), configHash); err != nil {
```
Comment (sebgl) [inline pkg/controller/agent/pod.go:194]: ugh 😬

File: pkg/controller/agent/pod.go [hunk 9]
```diff
@@ -50,63 +63,199 @@ var (
...
+					return nil, err
+				}
+			}
+		}
 	}
 
-	for i, association := range params.Agent.GetAssociations() {
+	if esAssociation != nil {
+		builder = builder.
+			WithCommand([]string{"/usr/bin/env", "bash", "-c", trustCAScript(path.Join(certificatesDir(esAssociation), CAFileName))})
```
Comment (sebgl) [inline pkg/controller/agent/pod.go:203]: that's the place where I'd move the long comment about the ca reload stuff

File: pkg/controller/agent/pod.go [hunk 10]
```diff
@@ -50,63 +63,199 @@ var (
...
+		if fs != nil {
+			esAssociation = association.GetAssociationOfType(fs.GetAssociations(), commonv1.ElasticsearchAssociationType)
+			if esAssociation != nil {
+				if params.Agent.Namespace != esAssociation.Associated().GetNamespace() {
+					return nil, fmt.Errorf(
+						"agent namespace %s is different than referenced Elasticsearch namespace %s, this is not supported yet",
+						params.Agent.Namespace,
+						esAssociation.Associated().GetNamespace(),
+					)
+				}
```
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/agent/pod.go:180]: As a general principle we try not to have controller need to fetch their referenced resource. The association controllers should do whatever's required so regular controllers have everything in place in the resource namespace.

File: pkg/controller/agent/pod.go [hunk 11]
```diff
@@ -50,63 +63,199 @@ var (
...
+	return &fs, nil
+}
 
-	return builder.PodTemplate
+func trustCAScript(caPath string) string {
+	return fmt.Sprintf(`#!/usr/bin/env bash
+set -e
+cp %s /etc/pki/ca-trust/source/anchors/
+update-ca-trust
+/usr/bin/tini -- /usr/local/bin/docker-entrypoint -e
```
Comment (sebgl) [inline pkg/controller/agent/pod.go:257]: I'm fine with that considering it's short term only... (otherwise I'd expect we don't override the container entrypoint in case things change in future versions of agent, and rather rely on init containers.)

**Review by sebgl — COMMENTED** [review]
There seems to be quite a few rough edges pending improvements in Agent itself.
Especially what concerns the communication with the Elasticsearch cluster where data should be sent (its URL is in... the Kibana config? but we also need a CA so we retrieve it transitively through the fleet server association from the Agent controller without an explicit association... which goes a bit against our usual association controller concept).

I'd be fine to merge as is though, considering the CRD is still alpha.

The multiple flavors of Agent (server vs. not, fleet vs. not) leak into many functions throughout the code but I'm not sure we can easily avoid this.

File: pkg/apis/agent/v1alpha1/validations.go [hunk 12]
```diff
@@ -104,10 +111,117 @@ func checkSingleConfigSource(b *Agent) field.ErrorList {
...
+				"remove config, it can't be set in fleet mode",
+			))
+		}
+
+		if a.Spec.ConfigRef != nil {
+			errors = append(errors, field.Invalid(
+				field.NewPath("spec").Child("configRef"),
+				a.Spec.ConfigRef,
+				"remove configRef, it can't be set in fleet mode",
+			))
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/apis/agent/v1alpha1/validations.go:139]: Elastic Agent config file is empty in case of fleet mode, there is nothing to override. I think it would be misguiding to allow specifying any config when it's not even read by anything.

File: pkg/controller/agent/pod.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/agent/pod.go:180]: Short-term, yes. Long term Kibana should push (through Fleet Server) everything to Elastic Agents, today it doesn't work as per "known limitations" in the PR description.

File: pkg/controller/agent/pod.go (hunk 7 above)
Comment (david-kow) [reply pkg/controller/agent/pod.go:190]: Because we include Fleets Server association to ES in Elastic Agents config hash.

File: pkg/controller/agent/pod.go (hunk 8 above)
Comment (david-kow) [reply pkg/controller/agent/pod.go:194]: Could you elaborate? I'm not sure how I can address this comment.

File: pkg/controller/agent/pod.go (hunk 11 above)
Comment (david-kow) [reply pkg/controller/agent/pod.go:257]: Yes, it's supposed to be short term.

File: pkg/controller/agent/pod.go (hunk 9 above)
Comment (david-kow) [reply pkg/controller/agent/pod.go:203]: To me it feels a bit off to have this comment at the end of the code that does the work, but I've moved it as requested.

File: pkg/apis/agent/v1alpha1/agent_types.go (hunk 6 above)
Comment (david-kow) [reply pkg/apis/agent/v1alpha1/agent_types.go:92]: I think this is a viable idea, but it would still require to define `elasticsearchRefs` outside of `fleet` field. This feels confusing as it is a required field for Fleet Server. It allows to tie together some of the fields, but users still need to know how it should be used with Fleet Server:
```yaml
elasticsearchRefs:
- name: es
fleet:
    managed: true
    server: true
    serverRef: {}
    kibanaRef: kb
```

and plain Elastic Agent:
```yaml
elasticsearchRefs: {}
fleet:
    managed: true
    server: false
    serverRef: fs
    kibanaRef: kb
```

Overall, I'm not sure if it's an improvement over keeping everything at the same level.

**Comment by david-kow** [comment]
> Sorry for being very late to this.
I think the mode: standalone | fleet_server | fleet_agent alternative you proposed in #4429 would make slightly more sense to me?

We chatted offline with @sebgl and agreed that this mixes two flags together (mode and whether Elastic Agent should run Fleet Server), so the current approach with separate flags is a better one.

**Review by barkbay — COMMENTED** [review]
I'm still doing some tests, quick feedback:

* I inadvertently set a bad setting for `xpack.fleet.agents.fleet_server.hosts`, I'm doing my tests in a namespace named `fleet-demo`, I should have set `xpack.fleet.agents.fleet_server.hosts` to `https://fleet-server-agent-http.fleet-demo.svc:8220`. Even once fixed and everything restarted (Kibana and Agents) I'm not able to get the agents connected, they keep using the wrong URL:

```
2021-07-13T07:43:52.718Z	WARN	[transport]	transport/tcp.go:52	DNS lookup failure "fleet-server-agent-http.default.svc": lookup fleet-server-agent-http.default.svc on 10.55.192.10:53: no such host
2021-07-13T07:43:52.718Z	ERROR	fleet/fleet_gateway.go:180	failed to dispatch actions, error: fail to communicate with updated API client hosts: Get "https://fleet-server-agent-http.default.svc:8220/api/status?": lookup fleet-server-agent-http.default.svc on 10.55.192.10:53: no such host
```

* When agents are restarted they get a new name (or uid, idk what is used internally), which generates a new item in the Kibana UI. Not sure how we can handle that, I guess we would have to use a volume to save the state somewhere ?

image

File: pkg/controller/agent/pod.go [hunk 13]
```diff
@@ -50,63 +63,200 @@ var (
...
+				caSecretName := esAssociation.AssociationConf().GetCASecretName()
+				vols = append(vols, volume.NewSecretVolumeWithMountPath(
+					caSecretName,
+					fmt.Sprintf("%s-certs", esAssociation.AssociationType()),
+					certificatesDir(esAssociation),
+				))
+
+				// Because of the reference chain (Elastic Agent ---> Fleet Server ---> Elasticsearch), we are going to get
+				// notified when CA of Elasticsearch changes as Fleet Server resource will get updated as well. But what we
+				// also need to do is to roll Elastic Agent Pods to pick up the update CA. To do be able to do that, we are
```
Comment (barkbay) [inline pkg/controller/agent/pod.go:187]:
```suggestion
				// also need to do is to roll Elastic Agent Pods to pick up the update CA. To be able to do that, we are
```

File: pkg/apis/agent/v1alpha1/agent_types.go [hunk 14]
```diff
@@ -126,6 +156,28 @@ const (
 	AgentGreenHealth AgentHealth = "green"
 )
 
+// +kubebuilder:validation:Enum=standalone;fleet
```
Comment (barkbay) [inline pkg/apis/agent/v1alpha1/agent_types.go:159]: This requires de regenerate the crds.

```
10:20:37  Error: dirty local changes
10:20:37   M config/crds/v1/all-crds.yaml
10:20:37   M config/crds/v1/bases/agent.k8s.elastic.co_agents.yaml
10:20:37   M config/crds/v1beta1/all-crds.yaml
10:20:37   M config/crds/v1beta1/bases/agent.k8s.elastic.co_agents.yaml
10:20:37   M deploy/eck-operator/charts/eck-operator-crds/templates/all-crds-legacy.yaml
10:20:37   M deploy/eck-operator/charts/eck-operator-crds/templates/all-crds.yaml
```

File: pkg/controller/association/conf.go [hunk 15]
```diff
@@ -157,6 +157,20 @@ func GetAssociationConf(association commonv1.Association) (*commonv1.Association
 	return extractAssociationConf(annotations, association.AssociationConfAnnotationName())
 }
 
+// GetAssociationOfType returns the first association from the provided slice that matches provided type. Returns
+// nil if such association can't be found.
+func GetAssociationOfType(
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/association/conf.go:162]: It is possible (Agent has multiple ES references), but not expected where this function is currently used. I've renamed it and made it return an error when multiple associations of the same type are found, as suggested.

**Comment by david-kow** [comment]
> * I inadvertently set a bad setting for `xpack.fleet.agents.fleet_server.hosts`, I'm doing my tests in a namespace named `fleet-demo`, I should have set `xpack.fleet.agents.fleet_server.hosts` to `https://fleet-server-agent-http.fleet-demo.svc:8220`. Even once fixed and everything restarted (Kibana and Agents) I'm not able to get the agents connected, they keep using the wrong URL.

That's a good catch. The bug seems to be on the Kibana side. After applying the fixed config, Kibana is still using the old one (as can be seen in the UI in Fleet > Settings flyout). I'll file a bug regarding this.


> * When agents are restarted they get a new name (or uid, idk what is used internally), which generates a new item in the Kibana UI. Not sure how we can handle that, I guess we would have to use a volume to save the state somewhere ? _(Edit: might be related to [Adjust Elastic Agent controller to make use of new container state handling in 7.13 #4428](https://github.com/elastic/cloud-on-k8s/issues/4428) ?)_

Yes, this is best we can currently do given the existing [issue](https://github.com/elastic/beats/issues/25326).

**Review by barkbay — COMMENTED** [review]
I'm having this error when running the e2e test `TestFleetMode`:

```
  - lastTransitionTime: "2021-07-13T09:47:49Z"
    lastUpdateTime: "2021-07-13T09:47:49Z"
    message: 'pods "test-agent-fleet-fs-c7hs-agent-7bdb4658ff-" is forbidden: unable
      to validate against any pod security policy: [spec.volumes[0]: Invalid value:
      "hostPath": hostPath volumes are not allowed to be used spec.containers[0].securityContext.runAsUser:
      Invalid value: 0: running with the root UID is forbidden]'
```

```
NAME                                                               HEALTH   NODES   VERSION   PHASE   AGE
elasticsearch.elasticsearch.k8s.elastic.co/test-agent-fleet-xkzq   green    3       7.13.1    Ready   7m19s

NAME                                                 HEALTH   NODES   VERSION   AGE
kibana.kibana.k8s.elastic.co/test-agent-fleet-7wfb   green    1       7.13.1    7m19s

NAME                                                  HEALTH   AVAILABLE   EXPECTED   VERSION   AGE
agent.agent.k8s.elastic.co/test-agent-fleet-ea-fc9h                                             7m19s
agent.agent.k8s.elastic.co/test-agent-fleet-fs-c7hs   red                                       7m19s

NAME                                            READY   STATUS    RESTARTS   AGE
pod/test-agent-fleet-7wfb-kb-7f77d7c7cf-rgdkv   1/1     Running   0          7m11s
pod/test-agent-fleet-xkzq-es-masterdata-0       1/1     Running   0          7m13s
pod/test-agent-fleet-xkzq-es-masterdata-1       1/1     Running   0          7m13s
pod/test-agent-fleet-xkzq-es-masterdata-2       1/1     Running   0          7m13s
```

I'm investigating if it comes from my setup.

**Comment by david-kow** [comment]
I think you don't have PSP (`config/recipes/psp/beats-agent-psp.yaml`) applied, it does contain permission for `hostPath`. Having said that, we shouldn't strictly require it for Fleet mode.

**Review by barkbay — APPROVED** [review]
LGTM ! 🚢

File: pkg/controller/agent/pod.go [hunk 16]
```diff
@@ -50,63 +63,200 @@ var (
...
+		// we need first to identify the Fleet Server and then identify its reference to Elasticsearch.
+		fs, err := getAssociatedFleetServer(params)
+		if err != nil {
+			return nil, err
+		}
+
+		if fs != nil {
+			esAssociation = association.GetAssociationOfType(fs.GetAssociations(), commonv1.ElasticsearchAssociationType)
+			if esAssociation != nil {
+				if params.Agent.Namespace != esAssociation.Associated().GetNamespace() {
```
[the remark this answers is not in the document]
Comment (david-kow) [reply pkg/controller/agent/pod.go:178]: This will come as a separate PR.
